File: Week_01/G20200343030499/LeetCode_189_449.java
/*
 * @lc app=leetcode id=189 lang=java
 *
 * [189] Rotate Array
 */

// 时间复杂度： O(n^2)
// 空间复杂度： O(1)
// @lc code=start
class Solution {
    public void rotate(int[] nums, int k) {
        while (k-- > 0) {
            moveToRightByOneStep(nums);
        }
    }

    private void moveToRightByOneStep(int[] nums) {
        int temp = nums[nums.length - 1];
        for (int i = nums.length - 2; i >= 0; i--) {
            nums[i + 1] = nums[i];
        }
        nums[0] = temp;
    }
}
// @lc code=end

File: Week_01/G20200343030449/LeetCode_189_449.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        if(nums.size()==0) return;

        auto move = k%nums.size();

        reverse(nums, 0, nums.size());
        reverse(nums, 0, move);
        reverse(nums, move, nums.size());
    }

    void reverse(vector<int>& nums, int start, int end) {
        auto i = start;
        auto j = end-1;

        while (i<j) {
            nums[i] = nums[i]^nums[j];
            nums[j] = nums[i]^nums[j];
            nums[i] = nums[i]^nums[j];
            i++,j--;
        }

        return;
    }
};
